mp/term.py

```diff
@@ -65,10 +69,132 @@ def set_rx_encoding(self, enc):
+                'A': '\1b[18~',  # F7
+                'B': '\1b[19~',  # F8
+                'C': '\1b[20~',  # F9
+                'D': '\1b[21~',  # F10
+            }
+            navcodes = {
+                'H': '\x1b[A',  # UP
+                'P': '\x1b[B',  # DOWN
+                'K': '\x1b[D',  # LEFT
+                'M': '\x1b[C',  # RIGHT
+                'G': '\x1b[H',  # HOME
+                'O': '\x1b[F',  # END
+                'R': '\x1b[2~',  # INSERT
+                'S': '\x1b[3~',  # DELETE
+                'I': '\x1b[5~',  # PGUP
+                'Q': '\x1b[6~',  # PGDN        
+            }
+            
+            def __init__(self):
+                super(Console, self).__init__()
+                self._saved_ocp = ctypes.windll.kernel32.GetConsoleOutputCP()
+                self._saved_icp = ctypes.windll.kernel32.GetConsoleCP()
+                ctypes.windll.kernel32.SetConsoleOutputCP(65001)
+                ctypes.windll.kernel32.SetConsoleCP(65001)
+                # ANSI handling available through SetConsoleMode since Windows 10 v1511 
+                # https://en.wikipedia.org/wiki/ANSI_escape_code#cite_note-win10th2-1
+                # if platform.release() == '10' and int(platform.version().split('.')[2]) > 10586:
+                ENABLE_VIRTUAL_TERMINAL_PROCESSING = 0x0004
+                import ctypes.wintypes as wintypes
+                if not hasattr(wintypes, 'LPDWORD'): # PY2
+                    wintypes.LPDWORD = ctypes.POINTER(wintypes.DWORD)
+                SetConsoleMode = ctypes.windll.kernel32.SetConsoleMode
+                GetConsoleMode = ctypes.windll.kernel32.GetConsoleMode
+                GetStdHandle = ctypes.windll.kernel32.GetStdHandle
+                mode = wintypes.DWORD()
+                GetConsoleMode(GetStdHandle(-11), ctypes.byref(mode))
+                if (mode.value & ENABLE_VIRTUAL_TERMINAL_PROCESSING) == 0:
+                    SetConsoleMode(GetStdHandle(-11), mode.value | ENABLE_VIRTUAL_TERMINAL_PROCESSING)
+                    self._saved_cm = mode
+                self.output = codecs.getwriter('UTF-8')(Out(sys.stdout.fileno()), 'replace')
+                # the change of the code page is not propagated to Python, manually fix it
+                sys.stderr = codecs.getwriter('UTF-8')(Out(sys.stderr.fileno()), 'replace')
+                sys.stdout = self.output
+                self.output.encoding = 'UTF-8'  # needed for input
+
+            def __del__(self):
+                ctypes.windll.kernel32.SetConsoleOutputCP(self._saved_ocp)
+                ctypes.windll.kernel32.SetConsoleCP(self._saved_icp)
+                try:
+                    ctypes.windll.kernel32.SetConsoleMode(ctypes.windll.kernel32.GetStdHandle(-11), self._saved_cm)
+                except AttributeError: # in case no _saved_cm
+                    pass
+
+            def getkey(self):
+                while True:
+                    z = msvcrt.getwch()
```

Comment:
Can you add a few comments here to explain what this does?

Reply:
you can see https://en.wikipedia.org/wiki/ANSI_escape_code#Windows, This is the reason for the difference in system coding，the code changed the encoding data stream to utf-8.
`fncodes` and `navcodes` is utf-8 to Special ANSI
Because the Win32 console did not support ANSI escape sequences at all.